src/figuretype/depot.c

```diff
@@ -0,0 +1,395 @@
+        return 0;
+    }
+
+    building *src = building_get(order->src_storage_id);
+    building *dst = building_get(order->dst_storage_id);
+    if (!src || !dst) {
+        return 0;
+    }
+
+    if (src->state != BUILDING_STATE_IN_USE || dst->state != BUILDING_STATE_IN_USE) {
+        return 0;
+    }
+
+    if (src->type != BUILDING_GRANARY && src->type != BUILDING_WAREHOUSE) {
+        return 0;
+    }
+    if (dst->type != BUILDING_GRANARY && dst->type != BUILDING_WAREHOUSE) {
+        return 0;
+    }
+
+    int src_amount = src->type == BUILDING_GRANARY ?
+        building_granary_resource_amount(order->resource_type, src) / RESOURCE_ONE_LOAD :
+        building_warehouse_get_amount(src, order->resource_type);
+    int dst_amount = dst->type == BUILDING_GRANARY ?
+        building_granary_resource_amount(order->resource_type, dst) / RESOURCE_ONE_LOAD :
+        building_warehouse_get_amount(dst, order->resource_type);
+    if (src_amount == 0) {
+        return 0;
+    }
+
+    switch (order->condition.condition_type) {
+        case ORDER_CONDITION_SOURCE_HAS_MORE_THAN:
+            return src_amount >= order->condition.threshold;
+        case ORDER_CONDITION_DESTINATION_HAS_LESS_THAN:
+            return dst_amount < order->condition.threshold;
+        default:
+            return 1;
+    }
+}
+
+static int is_current_state_still_valid_for_order(const figure *f, const order *order)
+{
+    building *destination = building_get(f->destination_building_id);
+    if (destination->state != BUILDING_STATE_IN_USE) {
+        return 0;
+    }
+
+    switch (f->action_state) {
+        case FIGURE_ACTION_239_DEPOT_CART_PUSHER_HEADING_TO_SOURCE:
+        case FIGURE_ACTION_240_DEPOT_CART_PUSHER_AT_SOURCE:
+        {
+            if (f->destination_building_id == order->src_storage_id) {
+                return 1;
+            } else {
+                return 0;
+            }
+            break;
+        }
+        case FIGURE_ACTION_241_DEPOT_CART_HEADING_TO_DESTINATION:
+        case FIGURE_ACTION_242_DEPOT_CART_PUSHER_AT_DESTINATION:
+        {
+            if (f->resource_id == order->resource_type
+                && f->destination_building_id == order->dst_storage_id
+                && f->loads_sold_or_carrying > 0) {
+                return 1;
+            } else {
+                return 0;
+            }
+            break;
+        }
+    }
+    return 1;
+}
+
+static void dump_goods(figure *f)
+{
+    if (f->loads_sold_or_carrying > 0) {
+        // TODO maybe add a dump goods animation?
+        f->loads_sold_or_carrying = 0;
+        f->resource_id = RESOURCE_NONE;
+        set_cart_graphic(f);
+    }
+}
+
+static int storage_remove_resource(building *b, int resource, int amount)
+{
+    if (b->type == BUILDING_GRANARY) {
+        // For granaries, only take full units.
+        int source_amount = b->resources[resource];
+        if (source_amount < RESOURCE_ONE_LOAD) {
+            return amount;
```

Comment:
Instead of returning the remaining amount, I think it would make more sense to return the amount actually removed.

I know the old code didn't do that, but we can improve things with the new one.

Reply:
Done.